apps/meeting/src/providers/VideoGridProvider/state.tsx

```diff
@@ -0,0 +1,510 @@
+            ? TargetDisplaySize.High
+            : targetDisplaySize
+        )
+      );
+    }
+
+    // Set the common tiles to low priority
+    if (type === 'common') {
+      videoPreferences.add(
+        new VideoPreference(attendeeId, 2, targetDisplaySize)
+      );
+    }
+  }
+  priorityBasedPolicy.chooseRemoteVideoSources(videoPreferences.build());
+};
+
+export function reducer(state: State, { type, payload }: Action): State {
+  const { gridState, attendeeStates, videoSourceState } = state;
+
+  switch (type) {
+    case VideoGridAction.ResetVideoGridState: {
+      return initialState;
+    }
+    case VideoGridAction.UpdateAttendeeStates: {
+      const { roster } = payload;
+
+      // Remove attendee that left the meeting
+      for (const attendeeId of Object.keys(attendeeStates)) {
+        if (!isContentShare(attendeeId) && !(attendeeId in roster)) {
+          delete attendeeStates[attendeeId];
+        }
+      }
+
+      // Add attendee that joined the meeting
+      for (const attendeeId of Object.keys(roster)) {
+        const name = roster[attendeeId]?.name || '';
+
+        if (attendeeId in attendeeStates) {
+          attendeeStates[attendeeId].name = name;
+        } else {
+          attendeeStates[attendeeId] = {
+            attendeeId,
+            name,
+            videoEnabled: false,
+            bandwidthConstrained: false,
+          } as AttendeeState;
+        }
+      }
+
+      // Ensure the state of `videoEnabled` in the racing condition of UpdateAttendeeStates and UpdateVideoSources
+      // To do: Merge all the update actions to one
+      for (const attendeeId of videoSourceState.cameraSources) {
+        if (attendeeId in attendeeStates) {
+          attendeeStates[attendeeId].videoEnabled = true;
+        }
+      }
+
+      return {
+        ...state,
+        attendeeStates,
+      };
+    }
+    case VideoGridAction.UpdateVideoSources: {
+      const { videoSources } = payload as { videoSources: VideoSource[] };
+      const cameraSources: string[] = [];
+      const videoSourceIds = videoSources.map(
+        videoSource => videoSource.attendee.attendeeId
+      );
+
+      // Reset the `videoEnabled` of all attendeeStates
+      for (const attendee of Object.values(attendeeStates)) {
+        attendee.videoEnabled = false;
+      }
+
+      // Remove content share from attendeeStates,
+      // if content share is not in video sources
+      for (const attendeeId of Object.keys(attendeeStates)) {
+        if (
+          isContentShare(attendeeId) &&
+          !videoSourceIds.includes(attendeeId)
+        ) {
+          delete attendeeStates[attendeeId];
+        }
+      }
+
+      // Update the `videoEnabled` according to video sources
+      for (const attendeeId of videoSourceIds) {
+        if (!(attendeeId in attendeeStates)) {
+          if (isContentShare(attendeeId)) {
+            attendeeStates[attendeeId] = {
+              attendeeId,
+              name: 'content share',
```

Comment:
Any reason for the change in name here or keeping it '' (empty) in the else block?

Reply:
There's a racing condition between `remoteVideoSourcesDidChange` and `AttendeeIdPresence`. Here I use the `useRosterState` hook, where we skip the content share. So here, whenever we find the content share source via `remoteVideoSourcesDidChange`, we assign it name 'content share'.(Cause `useRosterState` skips the content share, and there is no way to update it's name, check [here](https://github.com/aws/amazon-chime-sdk-component-library-react/blob/d1fc0f5312e8cc4f47ebceab030cbbd8bc7bdb90/src/providers/RosterProvider/index.tsx#L44-L47)). But for non-content-share video source their name can be updated via `AttendeeIdPresence` (`useRosterState`).

So it's better to use `audioVideo.realtimeSubscribeToAttendeeIdPresence(cb)` here instead of `useRosterState`, but I didn't realize this when I implement the project. I should improve the code in the future to make things simple.